### ISSUE TITLE
constraints problem while proving SR

### ISSUE THREAD

**Issue by francoisthire** [rlepigre/lambdapi#72]
On the following example, `lambdapi` fails to prove SR

```
type : Type.

def eta : type -> Type.

prod : A : type -> (eta A -> type) -> type.

lam : A : type -> B : (eta A -> type) -> (x:eta A -> eta (B x)) -> eta (prod A B).

def app : A : type -> B : (eta A -> type) -> eta (prod A B) -> x: eta A -> eta (B x).

[A,f] eta (prod A f) --> a:eta A -> eta (f a).

[A,B,f,a] app _ _ (lam _ _ f) a --> f a.
```

Using option `--debug j`, here is the debugging message of `lambdapi`

```
[subj] check_rule [eta (prod ?A ?f) → !a:eta <?A>, eta (<?f> a)]
[subj] LHS has type [Type]
[subj] check_rule [app ?#3 ?#2 (lam ?#1 ?#0 ?f) ?a → <?f> <?a>]
[subj] LHS has type [eta (?#2 ?a)]
[subj]   if [eta ?#1] = [eta ?#3]
[subj]   if [eta (?#0 a)] = [eta (?#2 a)]
[subj] RHS has type [eta (?#2 ?a)]
[subj]   if [eta (?#0 ?a)] = [eta (?#2 ?a)]
[subj]   if [eta ?#3] = [eta ?#1]
[lf.dk, 13:0-39] Error on command.
Cannot solve [eta (?#0 ?a)] ~ [eta (?#2 ?a)]
Unable to prove SR for rule [app ?#3 ?#2 (lam ?#1 ?#0 ?f) ?a → <?f> <?a>].
```

Clearly, there is an issue because the LHS generates the constraint that it cannot solve.

**Comment by rlepigre** [comment]
Actually no. I you look closely, the term above contains `a` (a free variable), and the constraint to solve contains `?a` which corresponds to the pattern variable.

Dedukti also fails on this example. Is there any justification as to why it should be accepted?

**Comment by fblanqui** [comment]
In this last example (with no rule for eta), we get:

11:44 ~/lambdapi-git ./lambdapi.native --debug j 72.dk 
[subj] check_rule [app ?#3 ?#2 (lam ?#1 ?#0 ?f) ?a → <?f> <?a>]
[subj] LHS has type [eta (?#2 ?a)]
[subj]   if [eta (prod ?#1 ?#0)] = [eta (prod ?#3 ?#2)]
[subj] RHS has type [eta (?#2 ?a)]
[subj]   if [eta (?#0 ?a)] = [eta (?#2 ?a)]
[subj]   if [eta ?#3] = [eta ?#1]
[72.dk, 11:0-39] Error on command.
Cannot solve [eta ?#3] ~ [eta ?#1]
Cannot solve [eta (?#0 ?a)] ~ [eta (?#2 ?a)]
Unable to prove SR for rule [app ?#3 ?#2 (lam ?#1 ?#0 ?f) ?a → <?f> <?a>].

We should normally be able to prove SR. From [subj]   if [eta (prod ?#1 ?#0)] = [eta (prod ?#3 ?#2)], we should deduce ?#1 = ?#3 and ?#0 = ?#2, hence solve the constraints of the RHS.

**Comment by GuillaumeGen** [comment]
It depends on what you mean by 

> with no rule for eta

If you did not declare it as injective, even if no rules are declared (right now) on this symbol, you should not be able to infer the convertibility of the arguments from the convertibility of the whole term.

Otherwise, you would be able to add later rules which break the correctness of Dedukti.

**Comment by fblanqui** [comment]
Issue solved in commit 5341527dbed8d67e6b93a4f443878c3ee8fd8f7f.

You need to write:
```
inj eta : type -> Type.
```
to declare eta as injective.
